src/entities/priceImpact/index.ts

```diff
@@ -382,52 +384,85 @@ export class PriceImpact {
-            poolId,
-            kind,
-            assetIn,
-            assetOut,
-            amount,
-            userData,
-            rpcUrl,
-            chainId,
-        });
-
-        // simulate swap in the reverse direction
-        const amountFinal = await doSingleSwapQuery({
-            poolId: poolId,
-            kind: kind,
-            assetIn: assetOut,
-            assetOut: assetIn,
-            amount: amountResult,
-            userData,
-            rpcUrl,
-            chainId,
-        });
+    static swap = async (
+        swapInput: SwapInput,
+        rpcUrl?: string,
+        block?: bigint,
+    ): Promise<PriceImpactAmount> => {
+        // simulate swapping to get amount out
+        const swap = new Swap(swapInput);
+        const swapQueryOutput = await swap.query(rpcUrl, block);
+
+        // build reverse swap with expected amount as given amount
+        let expectedAmountsOut: bigint[];
+        let expectedAmountsIn: bigint[];
+        if (swapQueryOutput.swapKind === SwapKind.GivenIn) {
+            expectedAmountsOut = swapQueryOutput.pathAmounts ?? [
+                swapQueryOutput.expectedAmountOut.amount,
+            ];
+            expectedAmountsIn = swapInput.paths.map((p) => p.inputAmountRaw);
+        } else {
+            expectedAmountsIn = swapQueryOutput.pathAmounts ?? [
+                swapQueryOutput.expectedAmountIn.amount,
+            ];
+            expectedAmountsOut = swapInput.paths.map((p) => p.outputAmountRaw);
+        }
+        const reverseSwapInput: SwapInput = {
+            ...swapInput,
+            paths: swapInput.paths.map((p, i) => ({
+                ...p,
+                tokens: [...p.tokens].reverse(),
+                pools: [...p.pools].reverse(),
+                outputAmountRaw: expectedAmountsIn[i],
+                inputAmountRaw: expectedAmountsOut[i],
+            })),
+        };
 
-        // calculate price impact using ABA method
-        const priceImpact = MathSol.divDownFixed(
-            abs(amount - amountFinal),
-            amount * 2n,
+        // simulate reverse swapping to get amount in
+        const reverseSwap = new Swap(reverseSwapInput);
+        const reverseSwapQueryOutput = await reverseSwap.query(rpcUrl, block);
+
+        // get relevant amounts for price impact calculation
+        const tokenAIndex =
+            swapInput.swapKind === SwapKind.GivenIn
+                ? 0 // first token of the path
+                : swapInput.paths[0].tokens.length - 1; // last token of the path
+        const tokenA = new Token(
+            swapInput.chainId,
+            swapInput.paths[0].tokens[tokenAIndex].address,
+            swapInput.paths[0].tokens[tokenAIndex].decimals,
         );
+        let initialA: TokenAmount;
+        let finalA: TokenAmount;
+        if ('expectedAmountOut' in reverseSwapQueryOutput) {
+            // givenIn case
+            initialA = TokenAmount.fromRawAmount(
```

Comment:
I think you can use swap.inputAmount here instead of mapping?

Reply:
✅